File: evalforge/runner.py
"""EvalForge runner: load an eval set, score cases, aggregate, gate on threshold.

This is the 'unit tests for prompts' engine. In CI, a failing threshold returns a
non-zero exit code, blocking the deploy -- eval-driven development as a real gate.
"""

from __future__ import annotations

import json
from collections.abc import Callable
from dataclasses import asdict, dataclass
from pathlib import Path

import yaml

from evalforge.scorers import get_scorer


@dataclass
class CaseResult:
    id: str
    scorer: str
    weight: float
    score: float
    passed: bool


@dataclass
class EvalReport:
    aggregate: float
    threshold: float
    passed: bool
    cases: list[CaseResult]

    def to_json(self) -> str:
        return json.dumps(
            {**asdict(self), "cases": [asdict(c) for c in self.cases]}, indent=2
        )


def run_eval(
    eval_set_path: str | Path,
    target: Callable[[str], str],
    threshold: float | None = None,
) -> EvalReport:
    spec = yaml.safe_load(Path(eval_set_path).read_text(encoding="utf-8"))
    threshold = threshold if threshold is not None else float(spec.get("threshold", 0.8))

    results: list[CaseResult] = []
    for case in spec["cases"]:
        output = target(case["input"])
        scorer = get_scorer(case["scorer"])
        kwargs = {k: case[k] for k in ("expected", "pattern") if k in case}
        score = float(scorer(output, **kwargs))
        weight = float(case.get("weight", 1.0))
        case_threshold = float(case.get("pass_score", 0.5))
        results.append(
            CaseResult(
                id=case["id"],
                scorer=case["scorer"],
                weight=weight,
                score=score,
                passed=score >= case_threshold,
            )
        )

    total_w = sum(r.weight for r in results) or 1.0
    aggregate = round(sum(r.score * r.weight for r in results) / total_w, 4)
    return EvalReport(
        aggregate=aggregate,
        threshold=threshold,
        passed=aggregate >= threshold,
        cases=results,
    )

File: java/src/main/java/com/evalforge/Runner.java
// EvalForge runner: load an eval set, score cases, aggregate, gate on threshold.
//
// This is the 'unit tests for prompts' engine. In CI, a failing threshold returns a
// non-zero exit code, blocking the deploy -- eval-driven development as a real gate.

package com.evalforge;

import java.io.IOException;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import org.yaml.snakeyaml.Yaml;

public final class Runner {

    private Runner() {}

    public record CaseResult(String id, String scorer, double weight, double score, boolean passed) {}

    public record EvalReport(double aggregate, double threshold, boolean passed, List<CaseResult> cases) {}

    public static EvalReport runEval(Path evalSetPath, Function<String, String> target) throws IOException {
        return runEval(evalSetPath, target, null);
    }

    @SuppressWarnings("unchecked")
    public static EvalReport runEval(Path evalSetPath, Function<String, String> target, Double threshold)
            throws IOException {
        String text = Files.readString(evalSetPath, StandardCharsets.UTF_8);
        Map<String, Object> spec = new Yaml().load(text);

        double effectiveThreshold = threshold != null
            ? threshold
            : (spec.containsKey("threshold") ? toDouble(spec.get("threshold")) : 0.8);

        List<CaseResult> results = new ArrayList<>();
        List<Map<String, Object>> cases = (List<Map<String, Object>>) spec.get("cases");
        for (Map<String, Object> c : cases) {
            String output = target.apply(str(c.get("input")));
            String scorer = str(c.get("scorer"));
            String arg = scorer.equals("regex") ? str(c.get("pattern")) : str(c.get("expected"));
            double score = Scorers.score(scorer, output, arg);
            double weight = c.containsKey("weight") ? toDouble(c.get("weight")) : 1.0;
            double caseThreshold = c.containsKey("pass_score") ? toDouble(c.get("pass_score")) : 0.5;
            results.add(new CaseResult(str(c.get("id")), scorer, weight, score, score >= caseThreshold));
        }

        double totalW = results.stream().mapToDouble(CaseResult::weight).sum();
        if (totalW == 0.0) totalW = 1.0;
        double weighted = results.stream().mapToDouble(r -> r.score() * r.weight()).sum();
        double aggregate = BigDecimal.valueOf(weighted / totalW).setScale(4, RoundingMode.HALF_EVEN).doubleValue();
        return new EvalReport(aggregate, effectiveThreshold, aggregate >= effectiveThreshold, results);
    }

    private static double toDouble(Object v) {
        if (v instanceof Number n) return n.doubleValue();
        return Double.parseDouble(v.toString());
    }

    private static String str(Object v) {
        return v == null ? "" : v.toString();
    }
}
